Cyclismo/libs/JTurbo/src/org/cowboycoders/turbotrainers/CourseTracker.java

```diff
@@ -57,81 +69,133 @@ public CourseTracker(List<LatLongAlt> coursePoints) {
     }
     distanceMarkers = distanceLocationMap.keySet().toArray(new Double[0]);
     Arrays.sort(distanceMarkers);
+    nearestLocation = distanceLocationMap.get(distanceMarkers[0]);
   }
 
   /**
-   * Returns the nearest location to the specified distance.
-   *
-   * FIXME: Two distance, locations: (25.0, locA), (30.0, locB), we are at 26.0
-   *
-   * In this case distance - key is not less than min point spacing, so we go to the next location,
-   * which is not the closest.
-   *
-   * Although this never returns the same location?
-   *
-   * @param distance is the distance travelled.
-   * @param location TODO
-   * @return the difference.
+   * Must be polled at frequency >= 1Hz
    */
-  public double getNearestLocation(final double distance, LatLongAlt location) {
-    Double key = null;
-    double delta = 0.0;
+  public void updateSpeed(double newSpeed) {
+    // we assume speed doesn't change much between updates
+    this.speed = newSpeed * Conversions.KM_PER_HOUR_TO_METRES_PER_SECOND;
+    lastTimeStamp = currentTimeStamp;
+    currentTimeStamp = System.nanoTime() / (Math.pow(10, 6));
+
+    distanceIntegrator.add(System.nanoTime() / (Math.pow(10, 9)), speed);
+    double distance = distanceIntegrator.getIntegral();
+    if (newSpeed < ZERO_CUTOFF) {
+      // assume haven't moved
+      return;
+    }
+
+    Double key;
+    Double nextKey;
+    double delta;
+
+    // loop gives us index of last marker passed
     for (int i = lastKnownDistanceMarkerIndex; i < distanceMarkers.length; i++) {
       key = distanceMarkers[i];
       lastKnownDistanceMarkerIndex = i;
       delta = distance - key;
-      if (delta < MIN_POINT_SPACING_M) {
+      if (delta < resolution) {
```

Comment:
Don't forget we can "snap" to a marker in the other direction i.e delta can be positive. So resolution determines when we should snap to the next marker. Not sure if that is desirable behaviour though... what do you reckon?